src/nvim/option.c

```diff
@@ -3822,8 +3865,25 @@ set_bool_option (
    * End of handling side effects for bool options.
    */
 
+  // after handling side effects, call autocommand
+
   options[opt_idx].flags |= P_WAS_SET;
 
+  if (!starting) {
+    char_u buf_old[2], buf_new[2], buf_type[7];
+    vim_snprintf((char *)buf_old, 2, "%d", old_value ? true: false);
+    vim_snprintf((char *)buf_new, 2, "%d", value ? true: false);
```

Comment:
Though no, you are right about this: there are default promotions performed when calling functions, and one of this default promotion is integer promotion which promotes every integer having rank less then `int` to `int` (or `unsigned`).

“Integer promotions” are in C99, 6.3.1.1, end of paragraph 2:

> If an int can represent all values of the original type, the value is converted to an int;
> otherwise, it is converted to an unsigned int. These are called the integer
> promotions.48) All other types are unchanged by the integer promotions.

Default promotions when using function calls are defined in C99 6.5.2.2, paragraph 6:

> If the expression that denotes the called function has a type that does not include a
> prototype, the integer promotions are performed on each argument, and arguments that
> have type float are promoted to double. These are called the default argument
> promotions. If the number of arguments does not equal the number of parameters, the
> behavior is undefined. If the function is defined with a type that includes a prototype, and
> either the prototype ends with an ellipsis (, ...) or the types of the arguments after
> promotion are not compatible with the types of the parameters, the behavior is undefined.
> If the function is defined with a type that does not include a prototype, and the types of
> the arguments after promotion are not compatible with those of the parameters after
> promotion, the behavior is undefined, except for the following cases:

and 7 it is stated that they apply to `...`:

> If the expression that denotes the called function has a type that does include a prototype,
> the arguments are implicitly converted, as if by assignment, to the types of the
> corresponding parameters, taking the type of each parameter to be the unqualified version
> of its declared type. The ellipsis notation in a function prototype declarator causes
> argument type conversion to stop after the last declared parameter. The default argument
> promotions are performed on trailing arguments.